snap-preload.c

```diff
@@ -76,12 +77,33 @@ int shm_unlink(const char *name) {
 
 // library init
 static void __attribute__ ((constructor)) init(void) {
-  SNAP_INSTANCE_NAME = secure_getenv("SNAP_INSTANCE_NAME");
   if (secure_getenv("SNAP_PRELOAD_DEBUG")) {
     DEBUG = 1;
   }
 
   SAVE_ORIGINAL_SYMBOL(setgroups);
   SAVE_ORIGINAL_SYMBOL(shm_open);
   SAVE_ORIGINAL_SYMBOL(shm_unlink);
-}
+
+  // Store the snap instance name in the heap because of https://bugs.launchpad.net/maas/+bug/2020427
+  SNAP_INSTANCE_NAME = malloc(SNAP_INSTANCE_NAME_MAX_LEN + 1);
+  if (SNAP_INSTANCE_NAME == NULL) {
+      log("Failed to allocate memory for SNAP_INSTANCE_NAME");
+      return;
+  }
+
+  char *env_value = secure_getenv("SNAP_INSTANCE_NAME");
+  if (env_value == NULL){
+    log("SNAP_INSTANCE_NAME is empty");
+    return;
+  }
+
+  int env_value_len = strlen(env_value);
+  if (env_value_len > SNAP_INSTANCE_NAME_MAX_LEN) {
+      log("SNAP_INSTANCE_NAME exceeds maximum length of %d chars", SNAP_INSTANCE_NAME_MAX_LEN);
+      return;
+  }
+
+  strncpy(SNAP_INSTANCE_NAME, env_value, env_value_len);
+  SNAP_INSTANCE_NAME[env_value_len] = '\0';
```

Comment:
I think you should be able to `malloc()` with the length of the value returned by `secure_getenv`, and just `strcpy` it.

Reply:
good suggestion :+1: